File: 2022/day-09/day-09.c
/* --- Advent of Code 2023 - Day 9: Rope Bridge --- */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LENGTH 256
#define MAX_POS 6000
#define NO_KNOTS 10

static int tail_pos_visited[MAX_POS][2];
static int tail_pos_visited_count = 0;

/* point struct */
typedef struct {
    int x;
    int y;
} point;

/* add position to tail_pos_visited */
void add_pos(point pos) {
    int pos_found = 0;
    for (int i = 0; i < tail_pos_visited_count; i++) {
        if (tail_pos_visited[i][0] == pos.x && tail_pos_visited[i][1] == pos.y) {
            pos_found = 1;
            break;
        }
    }

    if (!pos_found) {
        if (tail_pos_visited_count >= MAX_POS) {
            fprintf(stderr, "Buffer too small: %d\n", MAX_POS);
            exit(1);
        }

        tail_pos_visited[tail_pos_visited_count][0] = pos.x;
        tail_pos_visited[tail_pos_visited_count][1] = pos.y;
        tail_pos_visited_count++;
    }

    return;
}


/* solution part-1 */
int part_1(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];

    /* file opened successfully */
    if (fp != NULL) {
        /* initialize head and tail */
        point head = {0, 0};
        point tail = {0, 0};
        add_pos(tail);

        while (fgets(buffer, MAX_LENGTH - 1, fp) != NULL) {
            char direction = buffer[0];
            char steps = atoi(&(buffer[2]));

            /* grid mapping of directions */
            for (int i = 0; i < steps; i++) {
                switch (direction) {
                    case 'R':
                        head.x += 1;
                        break;
                    case 'L':
                        head.x -= 1;
                        break;
                    case 'U':
                        head.y += 1;
                        break;
                    case 'D':
                        head.y -= 1;
                        break;
                    default:
                        fprintf(stderr, "Invalid direction: '%c'\n", direction);
                        exit(1);
                }

                int diff_x = head.x - tail.x;
                int diff_y = head.y - tail.y;

                /* check if tail is touching head ... */
                if (abs(diff_x) <= 1 && abs(diff_y) <= 1) {
                    continue;

                /* ... otherwise adjust the tail with simple imlementation of signum function */
                } else {
                    tail.x += (diff_x > 0) - (diff_x < 0);
                    tail.y += (diff_y > 0) - (diff_y < 0);
                    add_pos(tail);
                }
            }
        }

        printf("%d\n", tail_pos_visited_count);

    /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* solution part-2 */
int part_2(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];

    /* file opened successfully */
    if (fp != NULL) {
        /* initialize head and tail */
        point knots[NO_KNOTS];
        for (int i = 0; i < NO_KNOTS; i++) {
            knots[i].x = 0;
            knots[i].y = 0;
        }
        add_pos(knots[9]);

        while (fgets(buffer, MAX_LENGTH - 1, fp) != NULL) {
            char direction = buffer[0];
            char steps = atoi(&(buffer[2]));

            /* grid mapping of directions */
            for (int i = 0; i < steps; i++) {
                switch (direction) {
                    case 'R':
                        knots[0].x += 1;
                        break;
                    case 'L':
                        knots[0].x -= 1;
                        break;
                    case 'U':
                        knots[0].y += 1;
                        break;
                    case 'D':
                        knots[0].y -= 1;
                        break;
                    default:
                        fprintf(stderr, "Invalid direction: '%c'\n", direction);
                        exit(1);
                }

                for (int i = 0; i < NO_KNOTS - 1; i++) {
                    int diff_x = knots[i].x - knots[i + 1].x;
                    int diff_y = knots[i].y - knots[i + 1].y;

                    /* check if tail is touching head ... */
                    if (abs(diff_x) <= 1 && abs(diff_y) <= 1) {
                        continue;

                    /* ... otherwise adjust the tail with simple imlementation of signum function */
                    } else {
                        knots[i + 1].x += (diff_x > 0) - (diff_x < 0);
                        knots[i + 1].y += (diff_y > 0) - (diff_y < 0);
                    }
                }
                add_pos(knots[9]);
            }
        }

        printf("%d\n", tail_pos_visited_count);

    /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* main function */
int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s -p[1|2] <input-file>\n", argv[0]);
        return 1;
    }

    if (!strcmp(argv[1], "-p1")) {
        part_1(argv[2]);
    } else if (!strcmp(argv[1], "-p2")) {
        part_2(argv[2]);
    } else {
        fprintf(stderr, "Invalid argument: '%s'\n", argv[1]);
        return 1;
    }

    return 0;
}

File: 2022/day-09/day-09.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" --- Advent of Code 2022 - Day 9: Rope Bridge --- """

import numpy as np
import sys


# solution part 1
def part_1():
    with open("input.txt") as f:
        # represent the position of head and tail as a tuple
        head = (0, 0)
        tail = (0, 0)

        # use a set as data structure, so that duplicates are ignored
        tail_pos_visited = set()
        # mark the starting position as visited
        tail_pos_visited.add(tail)

        for line in f.readlines():
            direction, steps = line.split()

            # grid mapping of directions
            for _ in range(int(steps)):
                match direction:
                    case 'R':
                        head = (head[0] + 1, head[1])
                    case 'L':
                        head = (head[0] - 1, head[1])
                    case 'U':
                        head = (head[0], head[1] + 1)
                    case 'D':
                        head = (head[0], head[1] - 1)

                diff_x = head[0] - tail[0]
                diff_y = head[1] - tail[1]

                # check if tail is touching head ...
                if abs(diff_x) <= 1 and abs(diff_y) <= 1:
                    continue

                # ... otherwise adjust the tail
                else:
                    tail = (tail[0] + np.sign(diff_x), tail[1] + np.sign(diff_y))
                    tail_pos_visited.add(tail)

    print(len(tail_pos_visited))


# solution part 2
def part_2():
    with open("input.txt") as f:
        # define amount of knots and store their position in array
        no_knots = 10
        knots = []

        # all 10 knots starting at (0,0)
        for _ in range(no_knots):
            knots.append((0, 0))

        # use a set as data structure, so that duplicates are ignored
        tail_pos_visited = set()
        # mark the starting position as visited
        tail_pos_visited.add(knots[-1])

        for line in f.readlines():
            direction, steps = line.split()

            # grid mapping for the directions
            for _ in range(int(steps)):
                match direction:
                    case 'R':
                        knots[0] = (knots[0][0] + 1, knots[0][1])
                    case 'L':
                        knots[0] = (knots[0][0] - 1, knots[0][1])
                    case 'U':
                        knots[0] = (knots[0][0], knots[0][1] + 1)
                    case 'D':
                        knots[0] = (knots[0][0], knots[0][1] - 1)

                for k in range(no_knots-1):
                    diff_x = knots[k][0] - knots[k + 1][0]
                    diff_y = knots[k][1] - knots[k + 1][1]

                    # check if knot #k is touching knot #k+1 ...
                    if abs(diff_x) <= 1 and abs(diff_y) <= 1:
                        continue

                    # ... otherwise adjust knot #k+1 and add position of last knot (index: knots[-1]) to visited positions
                    else:
                        knots[k + 1] = (knots[k + 1][0] + np.sign(diff_x), knots[k + 1][1] + np.sign(diff_y))
                        tail_pos_visited.add(knots[-1])

    print(len(tail_pos_visited))


# main entry point
if len(sys.argv) > 1 and sys.argv[1] == "-p1":
    part_1()
elif len(sys.argv) > 1 and sys.argv[1] == "-p2":
    part_2()
else:
    print("Usage: python", sys.argv[0], "[-p1|-p2]")
    sys.exit(1)
